### PR TITLE
Improved the detection of genome reusability

### PR DESCRIPTION
**Pull request by muffato**
It all started when we found out that ProteinTrees/CheckGenomedbReusability.pm was only comparing some properties of _protein-coding_ genes only to assess the reusabiity of genomes. I have fixed this and the query now includes (almost) all the fields that we use to make members.
Then I have rearranged the modules into a parent class and specific sub-modules for each pipeline (LoadMembers, ProteinTrees, EPO) so that they only test what matters.

The previous code caused two data bugs:
1. By only considering protein-coding genes, it was blind to updates to ncRNAs. Since some species have had their ncRNA-set updated recently, our current database is out of sync as it still includes the previous set of ncRNAs ... In e99, we will likely have to mark those species as "do_not_reuse" (although I'm also considering whether the e98 database could be patched instead).
2. The other issue is that our seq_member stable_ids for yeast _don't exist_. We still have the stable_ids that were in the e96 core db, but they have been renamed in the e97 core database ...

### CODE REVIEW

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckCanonMembersReusability.pm [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use strict;
+use warnings;
+
+use base ('Bio::EnsEMBL::Compara::RunnableDB::CheckGenomeReusability');
+
+
+sub run_comparison {
+    my $self = shift @_;
+
+    return $self->do_comparison('members',
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckCanonMembersReusability.pm]:
```suggestion
    return $self->do_one_comparison('members',
```

File: modules/Bio/EnsEMBL/Compara/PipeConfig/LoadMembers_conf.pm [hunk 2]
```diff
@@ -252,7 +252,7 @@ sub pipeline_analyses {
 # ---------------------------------------------[filter genome_db entries into reusable and non-reusable ones]------------------------
 
         {   -logic_name => 'check_reusability',
-            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::ProteinTrees::CheckGenomedbReusability',
+            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::ProteinTrees::CheckGenomeReusability',
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/LoadMembers_conf.pm]:
```suggestion
            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::CheckGenomeReusability',
```

File: modules/Bio/EnsEMBL/Compara/PipeConfig/MercatorPecan_conf.pm [hunk 3]
```diff
@@ -243,7 +243,7 @@ sub pipeline_analyses {
 # ---------------------------------------------[filter genome_db entries into reusable and non-reusable ones]------------------------
 
         {   -logic_name => 'check_reusability',
-            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::ProteinTrees::CheckGenomedbReusability',
+            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::ProteinTrees::CheckGenomeReusability',
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/MercatorPecan_conf.pm]:
```suggestion
            -module     => 'Bio::EnsEMBL::Compara::RunnableDB::CheckGenomeReusability',
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckMembersReusability.pm [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        'needs_core_db' => 1,
+    }
+}
+
+
+sub run_comparison {
+    my $self = shift @_;
+
+    if (comes_from_core_database($self->param('genome_db'))) {
+        return $self->do_comparison('exons',
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckMembersReusability.pm]:
```suggestion
        return $self->do_one_comparison('exons',
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckMembersReusability.pm [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            $self->hash_all_exons_from_dba( $self->param('prev_core_dba') ),
+            $self->hash_all_exons_from_dba( $self->param('curr_core_dba') ),
+        );
+    } else {
+        return $self->do_comparison('sequences',
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/ProteinTrees/CheckMembersReusability.pm]:
```suggestion
        return $self->do_one_comparison('sequences',
```

File: modules/Bio/EnsEMBL/Compara/PipeConfig/LoadMembers_conf.pm (hunk 2 above)
Comment (muffato) [reply modules/Bio/EnsEMBL/Compara/PipeConfig/LoadMembers_conf.pm]: Wrong module indeed, but it's `Bio::EnsEMBL::Compara::RunnableDB::ProteinTrees::CheckMembersReusability` that should be used

File: modules/Bio/EnsEMBL/Compara/PipeConfig/MercatorPecan_conf.pm (hunk 3 above)
Comment (muffato) [reply modules/Bio/EnsEMBL/Compara/PipeConfig/MercatorPecan_conf.pm]: same here

File: modules/Bio/EnsEMBL/Compara/RunnableDB/CheckGenomeReusability.pm [hunk 6]
```diff
@@ -197,23 +189,7 @@ sub run {
...
-    my ($removed, $remained1) = check_hash_equals($prev_hash, $curr_hash);
-    my ($added, $remained2)   = check_hash_equals($curr_hash, $prev_hash);
-
-    my $coding_objects_differ = $added || $removed;
-    if($coding_objects_differ) {
-        $self->warning("The coding objects changed: $added hash keys were added and $removed were removed");
-    }
-
-    $self->param('reuse_this', $coding_objects_differ ? 0 : 1);
+    $self->param('reuse_this', $self->run_comparison());
```
Comment (JAlvarezJarreta) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/CheckGenomeReusability.pm]: Maybe this method could be added in a comment somehow so everybody knows it needs to be implemented by each subclass?

**Pull request merged**